File: WitchcraftApp/src/test/java/com/tarbadev/witchcraft/DatabaseRecipeRepositoryTest.java
package com.tarbadev.witchcraft;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.jdbc.AutoConfigureTestDatabase;
import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest;
import org.springframework.boot.test.autoconfigure.orm.jpa.TestEntityManager;
import org.springframework.test.annotation.DirtiesContext;
import org.springframework.test.context.ActiveProfiles;
import org.springframework.test.context.junit4.SpringRunner;

import java.util.Arrays;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;

@RunWith(SpringRunner.class)
@DataJpaTest
@AutoConfigureTestDatabase(replace = AutoConfigureTestDatabase.Replace.NONE)
@DirtiesContext(classMode = DirtiesContext.ClassMode.BEFORE_EACH_TEST_METHOD)
@ActiveProfiles("test")
public class DatabaseRecipeRepositoryTest {
  @Autowired private TestEntityManager entityManager;
  @Autowired private RecipeRepository recipeRepository;

  private DatabaseRecipeRepository subject;

  @Before
  public void setUp() {
    subject = new DatabaseRecipeRepository(recipeRepository);
  }

  @Test
  public void createRecipe_ReturnsRecipe() {
    String recipe_url = "URL";

    Recipe recipe = Recipe.builder()
        .url(recipe_url)
        .ingredients(Collections.emptyList())
        .build();
    recipe = subject.createRecipe(recipe);
    Recipe expectedRecipe = Recipe.builder()
        .id(recipe.getId())
        .url(recipe_url)
        .ingredients(Collections.emptyList())
        .build();

    assertThat(recipe).isEqualToComparingFieldByFieldRecursively(expectedRecipe);
  }

  @Test
  public void createRecipe_savesIngredients() {
    Recipe recipe = Recipe.builder()
        .ingredients(Arrays.asList(
            Ingredient.builder().build(),
            Ingredient.builder().build()
        ))
        .build();

    Recipe returnedRecipe = subject.createRecipe(recipe);

    Recipe expectedRecipe = Recipe.builder()
        .id(returnedRecipe.getId())
        .ingredients(Arrays.asList(
            Ingredient.builder()
                .id(returnedRecipe.getIngredients().get(0).getId())
                .build(),
            Ingredient.builder()
                .id(returnedRecipe.getIngredients().get(1).getId())
                .build()
        ))
        .build();

    assertThat(returnedRecipe).isEqualToComparingFieldByFieldRecursively(expectedRecipe);
  }

  @Test
  public void findAll() {
    String url1 = "URL1";
    String url2 = "URL2";


    List<Recipe> expectedRecipes = Arrays.asList(
        entityManager.persist(Recipe.builder()
            .ingredients(Collections.emptyList())
            .url(url1)
            .imgUrl("imgUrl1")
            .build()),
        entityManager.persist(Recipe.builder()
            .ingredients(Collections.emptyList())
            .url(url2)
            .imgUrl("imgUrl2")
            .build())
    );

    entityManager.flush();
    entityManager.clear();

    assertThat(subject.findAll().size()).isEqualTo(expectedRecipes.size());
  }

  @Test
  public void findAll_returnsRecipeOrderedByName() {
    Recipe tartiflette = entityManager.persist(Recipe.builder().name("Tartiflette").ingredients(Collections.emptyList()).build());
    Recipe pizza = entityManager.persist(Recipe.builder().name("Pizza").ingredients(Collections.emptyList()).build());
    Recipe burger = entityManager.persistAndFlush(Recipe.builder().name("Burger").ingredients(Collections.emptyList()).build());

    entityManager.clear();

    List<Recipe> expectedRecipes = Arrays.asList(burger, pizza, tartiflette);

    List<Recipe> recipes = subject.findAll();
    assertThat(recipes.size()).isEqualTo(expectedRecipes.size());
    assertThat(recipes).isSortedAccordingTo(Comparator.comparing(Recipe::getName));
    assertThat(recipes.get(0).getName()).isEqualTo(burger.getName());
    assertThat(recipes.get(2).getName()).isEqualTo(tartiflette.getName());
  }

  @Test
  public void findById() {
    Recipe recipe = entityManager.persistAndFlush(Recipe.builder()
        .name("Recipe 1")
        .ingredients(Collections.emptyList())
        .steps(Collections.emptyList())
        .url("URL")
        .build()
    );

    entityManager.clear();

    assertThat(subject.findById(recipe.getId())).isEqualToComparingFieldByFieldRecursively(recipe);
  }

  @Test
  public void findById_returnsIngredientsOrderedByName() {
    Recipe recipe = entityManager.persistAndFlush(Recipe.builder()
        .name("Recipe 1")
        .ingredients(Arrays.asList(
            Ingredient.builder().name("Parsley").build(),
            Ingredient.builder().name("Cilantro").build(),
            Ingredient.builder().name("Egg").build()
        ))
        .steps(Collections.emptyList())
        .url("URL")
        .build()
    );

    Recipe expectedRecipe = recipe;
    expectedRecipe.getIngredients().sort(Comparator.comparing(Ingredient::getName));

    entityManager.clear();

    assertThat(subject.findById(recipe.getId())).isEqualToComparingFieldByFieldRecursively(expectedRecipe);
  }
}

File: WitchcraftApp/src/test/kotlin/com/tarbadev/witchcraft/recipes/persistence/repository/DatabaseRecipeRepositoryTest.kt
package com.tarbadev.witchcraft.recipes.persistence.repository

import com.tarbadev.witchcraft.converter.liter
import com.tarbadev.witchcraft.converter.unit
import com.tarbadev.witchcraft.recipes.domain.entity.Ingredient
import com.tarbadev.witchcraft.recipes.domain.entity.Recipe
import com.tarbadev.witchcraft.recipes.domain.entity.Step
import com.tarbadev.witchcraft.recipes.domain.entity.StepNote
import com.tarbadev.witchcraft.recipes.persistence.entity.IngredientEntity
import com.tarbadev.witchcraft.recipes.persistence.entity.RecipeEntity
import com.tarbadev.witchcraft.recipes.persistence.entity.StepEntity
import com.tarbadev.witchcraft.recipes.persistence.entity.StepNoteEntity
import org.hibernate.validator.internal.util.CollectionHelper.asSet
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.jdbc.AutoConfigureTestDatabase
import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest
import org.springframework.boot.test.autoconfigure.orm.jpa.TestEntityManager
import org.springframework.test.context.junit.jupiter.SpringExtension
import java.util.Collections.emptyList

@ExtendWith(SpringExtension::class)
@DataJpaTest
@AutoConfigureTestDatabase(replace = AutoConfigureTestDatabase.Replace.NONE)
class DatabaseRecipeRepositoryTest(
    @Autowired private val recipeEntityRepository: RecipeEntityRepository,
    @Autowired private val entityManager: TestEntityManager
) {
  private lateinit var databaseRecipeRepository: DatabaseRecipeRepository

  @BeforeEach
  fun setUp() {
    recipeEntityRepository.deleteAll()

    databaseRecipeRepository = DatabaseRecipeRepository(recipeEntityRepository)
  }

  @Test
  fun saveRecipe() {
    val recipeUrl = "URL"

    val recipe = databaseRecipeRepository.save(
        Recipe(
            name = "Lasagna",
            originUrl = recipeUrl,
            ingredients = emptyList(),
            steps = listOf(Step(name = "Add pasta", note = StepNote(comment = "Some note comment")))
        )
    )
    val savedStep = recipe.steps[0]
    val expectedRecipe = Recipe(
        id = recipe.id,
        name = "lasagna",
        originUrl = recipeUrl,
        ingredients = emptyList(),
        steps = listOf(Step(
            id = savedStep.id,
            name = "Add pasta",
            note = StepNote(id = savedStep.note.id, comment = "Some note comment")))
    )

    assertEquals(expectedRecipe, recipe)
  }

  @Test
  fun saveRecipe_savesIngredients() {
    val recipe = Recipe(
        name = "Lasagna",
        ingredients = listOf(
            Ingredient(quantity = 1.unit),
            Ingredient(quantity = 1.unit)
        ),
        steps = emptyList()
    )

    val returnedRecipe = databaseRecipeRepository.save(recipe)

    val expectedRecipe = Recipe(
        id = returnedRecipe.id,
        name = "lasagna",
        ingredients = listOf(
            Ingredient(id = returnedRecipe.ingredients[0].id, quantity = 1.unit),
            Ingredient(id = returnedRecipe.ingredients[1].id, quantity = 1.unit)
        ),
        steps = emptyList()
    )

    assertEquals(expectedRecipe, returnedRecipe)
  }

  @Test
  fun updateRecipe_updatesRecipe() {
    var recipe = databaseRecipeRepository.save(
        Recipe(
            name = "Name uncorrect",
            originUrl = "URL",
            ingredients = emptyList(),
            steps = emptyList()
        )
    )

    entityManager.clear()

    val modifiedRecipe = Recipe(
        id = recipe.id,
        name = "fixed name",
        originUrl = recipe.originUrl,
        imgUrl = recipe.imgUrl,
        ingredients = recipe.ingredients,
        steps = recipe.steps
    )

    recipe = databaseRecipeRepository.update(modifiedRecipe)

    assertEquals(modifiedRecipe.name, recipe.name)
  }

  @Test
  fun findAll() {
    val url1 = "URL1"
    val url2 = "URL2"


    val expectedRecipes = listOf(
        entityManager.persist(RecipeEntity(
            name = "Lasagna",
            ingredients = emptyList(),
            originUrl = url1,
            imgUrl = "imgUrl1"
        )),
        entityManager.persist(RecipeEntity(
            name = "Tartiflette",
            ingredients = emptyList(),
            originUrl = url2,
            imgUrl = "imgUrl2"
        ))
    )

    entityManager.flush()
    entityManager.clear()

    assertEquals(expectedRecipes.size, databaseRecipeRepository.findAll().size)
  }

  @Test
  fun findAll_returnsRecipeOrderedByName() {
    val tartiflette = toDomain(
        entityManager.persist(RecipeEntity(
            name = "Tartiflette",
            ingredients = emptyList(),
            steps = emptySet()
        ))
    )
    val pizza = toDomain(
        entityManager.persist(RecipeEntity(
            name = "Pizza",
            ingredients = emptyList(),
            steps = emptySet()
        ))
    )
    val burger = toDomain(
        entityManager.persistAndFlush(RecipeEntity(
            name = "Burger",
            ingredients = emptyList(),
            steps = emptySet()
        ))
    )

    entityManager.clear()

    val expectedRecipes = listOf(burger, pizza, tartiflette)

    val recipes = databaseRecipeRepository.findAll()
    assertEquals(expectedRecipes.size, recipes.size)
    assertEquals(burger.name, recipes[0].name)
    assertEquals(pizza.name, recipes[1].name)
    assertEquals(tartiflette.name, recipes[2].name)
  }

  @Test
  fun findAll_returnsOnlyNonArchivedRecipes() {
    toDomain(
        entityManager.persist(RecipeEntity(
            name = "Tartiflette",
            isArchived = true
        ))
    )
    val pizza = toDomain(
        entityManager.persist(RecipeEntity(
            name = "Pizza",
            isArchived = false
        ))
    )
    toDomain(
        entityManager.persistAndFlush(RecipeEntity(
            name = "Burger",
            isArchived = true
        ))
    )

    entityManager.clear()

    val expectedRecipes = listOf(pizza)

    val recipes = databaseRecipeRepository.findAll()
    assertEquals(expectedRecipes, recipes)
  }

  @Test
  fun findById() {
    val recipe = toDomain(
        entityManager.persistAndFlush(RecipeEntity(
            name = "Recipe 1",
            ingredients = emptyList(),
            steps = asSet(StepEntity(name = "First step", notes = listOf(StepNoteEntity(comment = "Some comment")))),
            originUrl = "URL"
        ))
    )

    entityManager.clear()

    assertEquals(recipe, databaseRecipeRepository.findById(recipe.id))
  }

  @Test
  fun findById_returnsIngredientsOrderedByName() {
    val recipe = entityManager.persistAndFlush(
        RecipeEntity(
            name = "Recipe 1",
            ingredients = listOf(
                IngredientEntity(name = "Parsley"),
                IngredientEntity(name = "Cilantro"),
                IngredientEntity(name = "Egg")
            ),
            steps = emptySet(),
            originUrl = "URL"
        )
    ).toRecipe()

    entityManager.clear()

    assertEquals(
        recipe.copy(ingredients = recipe.ingredients.sortedBy { it.name }),
        databaseRecipeRepository.findById(recipe.id)
    )
  }

  @Test
  fun findById_returnsNullWhenRecipeIsArchived() {
    val recipe = entityManager.persistAndFlush(
        RecipeEntity(
            name = "Recipe 1",
            isArchived = true
        )
    ).toRecipe()

    entityManager.clear()

    assertNull(databaseRecipeRepository.findById(recipe.id))
  }

  @Test
  fun delete() {
    val recipeEntity = entityManager.persistAndFlush(RecipeEntity(name = "Lasagna"))

    entityManager.clear()

    assertFalse(entityManager.find(RecipeEntity::class.java, recipeEntity.id)!!.isArchived)

    databaseRecipeRepository.delete(recipeEntity.id)

    entityManager.clear()

    assertTrue(entityManager.find(RecipeEntity::class.java, recipeEntity.id)!!.isArchived)
  }

  @Test
  fun setFavorite() {
    val recipe = entityManager.persistAndFlush(RecipeEntity(name = "Lasagna"))
    entityManager.clear()

    assertFalse(databaseRecipeRepository.findById(recipe.id)!!.favorite)

    databaseRecipeRepository.setFavorite(recipe.id, true)
    entityManager.clear()

    assertTrue(databaseRecipeRepository.findById(recipe.id)!!.favorite)
  }

  @Test
  fun findAllFavorite() {
    val recipes = listOf(
        entityManager.persist(RecipeEntity(favorite = true)),
        entityManager.persist(RecipeEntity(favorite = true)),
        entityManager.persist(RecipeEntity(favorite = false)),
        entityManager.persist(RecipeEntity(favorite = true)),
        entityManager.persist(RecipeEntity(favorite = true)),
        entityManager.persist(RecipeEntity(favorite = false)),
        entityManager.persist(RecipeEntity(favorite = true)),
        entityManager.persist(RecipeEntity(favorite = false))
    )
        .filter { it.favorite }
        .map { it.toRecipe() }
        .sortedBy { it.id }

    entityManager.flush()
    entityManager.clear()

    assertEquals(recipes, databaseRecipeRepository.findAllFavorite().sortedBy { it.id })
  }

  @Test
  fun findAllFavorite_returnsNonArchivedRecipes() {
    val recipes = listOf(
        entityManager.persist(RecipeEntity(favorite = true, isArchived = false)),
        entityManager.persist(RecipeEntity(favorite = true, isArchived = false)),
        entityManager.persist(RecipeEntity(favorite = false, isArchived = false)),
        entityManager.persist(RecipeEntity(favorite = true, isArchived = true)),
        entityManager.persist(RecipeEntity(favorite = true, isArchived = true)),
        entityManager.persist(RecipeEntity(favorite = false, isArchived = false)),
        entityManager.persist(RecipeEntity(favorite = true, isArchived = false)),
        entityManager.persist(RecipeEntity(favorite = false, isArchived = false))
    )
        .filter { it.favorite && !it.isArchived }
        .map { it.toRecipe() }
        .sortedBy { it.id }

    entityManager.flush()
    entityManager.clear()

    assertEquals(recipes, databaseRecipeRepository.findAllFavorite().sortedBy { it.id })
  }

  @Test
  fun findLastAddedRecipes() {
    val recipes = listOf(
        entityManager.persist(RecipeEntity(name = "", ingredients = emptyList(), steps = emptySet())),
        entityManager.persist(RecipeEntity(name = "", ingredients = emptyList(), steps = emptySet())),
        entityManager.persist(RecipeEntity(name = "", ingredients = emptyList(), steps = emptySet())),
        entityManager.persist(RecipeEntity(name = "", ingredients = emptyList(), steps = emptySet())),
        entityManager.persist(RecipeEntity(name = "", ingredients = emptyList(), steps = emptySet())),
        entityManager.persist(RecipeEntity(name = "", ingredients = emptyList(), steps = emptySet())),
        entityManager.persist(RecipeEntity(name = "", ingredients = emptyList(), steps = emptySet())),
        entityManager.persist(RecipeEntity(name = "", ingredients = emptyList(), steps = emptySet())),
        entityManager.persist(RecipeEntity(name = "", ingredients = emptyList(), steps = emptySet())),
        entityManager.persist(RecipeEntity(name = "", ingredients = emptyList(), steps = emptySet()))
    )
        .map { it.toRecipe() }
        .sortedBy { it.id }
        .reversed()
        .subList(0, 8)

    entityManager.flush()
    entityManager.clear()

    assertEquals(recipes, databaseRecipeRepository.findLastAddedRecipes())
  }

  @Test
  fun findLastAddedRecipes_returnsNonArchivedRecipes() {
    val recipes = listOf(
        entityManager.persist(RecipeEntity(isArchived = true)),
        entityManager.persist(RecipeEntity(isArchived = false)),
        entityManager.persist(RecipeEntity(isArchived = false)),
        entityManager.persist(RecipeEntity(isArchived = false))
    )

    entityManager.flush()
    entityManager.clear()

    assertEquals(recipes.size - 1, databaseRecipeRepository.findLastAddedRecipes().size)
  }

  @Test
  fun exists_returnsTrueWhenRecipeExists() {
    val recipeEntity = entityManager.persistAndFlush(
        RecipeEntity()
    )

    assertTrue(databaseRecipeRepository.existsById(recipeEntity.id))
  }

  @Test
  fun exists_returnsFalseWhenRecipeDoesNotExist() {
    assertFalse(databaseRecipeRepository.existsById(32))
  }

  private fun toDomain(recipeEntity: RecipeEntity) =
      Recipe(
          id = recipeEntity.id,
          name = recipeEntity.name.toLowerCase(),
          originUrl = recipeEntity.originUrl,
          imgUrl = recipeEntity.imgUrl,
          ingredients = recipeEntity.ingredients
              .map { ingredientEntity ->
                Ingredient(
                    id = ingredientEntity.id,
                    name = ingredientEntity.name,
                    quantity = ingredientEntity.quantity.liter
                )
              },
          steps = recipeEntity.steps
              .map { stepEntity ->
                Step(
                    id = stepEntity.id,
                    name = stepEntity.name,
                    note = stepEntity.notes.elementAtOrNull(0)?.toStepNote() ?: StepNote()
                )
              }
      )
}
